File: 12.整数转罗马数字.py
#
# @lc app=leetcode.cn id=12 lang=python3
#
# [12] 整数转罗马数字
#

# @lc code=start
class Solution:
    _rules = (
            ('M', 1000),
            ('CM', 900), ('D', 500), ('CD', 400), ('C', 100),
            ('XC', 90), ('L', 50), ('XL', 40), ('X', 10),
            ('IX', 9), ('V', 5), ('IV', 4), ('I', 1),
        )

    def intToRoman(self, num: int) -> str:
        """Enumerates cases in `_rules`. Complexity O(1). Avoids if-else branches."""
        result = []

        for roman, dec in Solution._rules:
            if num >= dec:
                count, num = divmod(num, dec)
                result.extend([roman] * count)

        return ''.join(result)

    """
    Original solution:

    DICT1 = {
        0: 'I',
        1: 'X',
        2: 'C',
        3: 'M'
    }

    DICT5 = {
        0: 'V',
        1: 'L',
        2: 'D',
    }
    def intToRoman(self, num: int) -> str:
        if num < 1 or num > 3999:
            return ''
        nstr = '%d' % num
        ans = ''
        for i, c in enumerate(nstr[::-1]):
            if c == '0':
                continue
            ci = int(c)
            if ci < 4:
                ans = Solution.DICT1[i]*ci + ans
            elif ci == 4:
                ans = Solution.DICT1[i] + Solution.DICT5[i] + ans
            elif ci < 9:
                ans = Solution.DICT5[i] + Solution.DICT1[i]*(ci-5) + ans
            else:   # ci == 9
                ans = Solution.DICT1[i] + Solution.DICT1[i+1] + ans
        return ans
    """

    
# @lc code=end
s = Solution()
print(s.intToRoman(1994))

File: 12.整数转罗马数字.java
import java.util.Map;
import java.util.LinkedHashMap;

/*
 * @lc app=leetcode.cn id=12 lang=java
 *
 * [12] 整数转罗马数字
 */

// @lc code=start
class Solution {
    private static final Map<Integer, String> RULES = new LinkedHashMap<>(){{
        put(1000, "M");
        put(900, "CM"); 
        put(500, "D"); 
        put(400, "CD");
        put(100, "C");
        put(90, "XC"); 
        put(50, "L");
        put(40, "XL"); 
        put(10, "X");
        put(9, "IX"); 
        put(5, "V");
        put(4, "IV"); 
        put(1, "I");
    }};

    public String intToRoman(int num) {
        StringBuffer ans = new StringBuffer();
        for (Map.Entry<Integer, String> e : RULES.entrySet()) {
            int dec = e.getKey();
            if (num >= dec) {
                int cnt = num / dec;
                num %= dec;
                for (int i = 0; i < cnt; ++i) {
                    ans.append(e.getValue());
                }
            }
        }
        return ans.toString();
    }
}
// @lc code=end

